### PR TITLE
Update Workflow To Allow Snyk test on Forked PRs

### PR DESCRIPTION
**Pull request by eslutz**
## Description

Updated the workflow trigger event to run in the context of the RockLib org so the workflows gets access to the org secrets, including the `SNYK_TOKEN` which is needed fro Snyk test to work.  Then the code from the pull request is pulled, built, and tested.

Unfortunately, all the reading I found on the topic warn about using this solution as it can allow untrusted code to execute using the org secrets.  However, we already have the suggested security measures in place to counteract this.  All pull requests from a forked repo require manual approval to run any workflow.  So be sure not to approve the workflow to run before reviewing the code.

This was the only solution I could find that would allow Snyk test to run on all pull requests.

[Using Environment Protection Rules to Secure Secrets When Building External Forks with pull_request_target](https://dev.to/petrsvihlik/using-environment-protection-rules-to-secure-secrets-when-building-external-forks-with-pullrequesttarget-hci)

## Type of change: 2

1. Non-functional change (e.g. documentation changes, removing unused `using` directives, renaming local variables, etc)
2. Bug fix (non-breaking change that fixes an issue)
3. New feature (non-breaking change that adds functionality)
4. Breaking change (fix or feature that could cause existing functionality to not work as expected)

## Checklist:

- Have you reviewed your own code? Do you understand every change?
- Are you following the [contributing guidelines](../blob/main/CONTRIBUTING.md)?
- Have you added tests that prove your fix is effective or that this feature works?
- New and existing unit tests pass locally with these changes?
- Have you made corresponding changes to the documentation?
- Will this change require an update to an example project? (if so, create an issue and link to it)

---

<sup>_[Reviewer guidelines](../blob/main/CONTRIBUTING.md#reviewing-changes)_</sup>

### CODE REVIEW

**Review by aatchison14 — DISMISSED** [review]

File: .github/workflows/build-and-test.yml [hunk 1]
```diff
@@ -21,9 +21,11 @@ jobs:
       SNYK_EXCLUSIONS: ${{ steps.snyk-exclusions.outputs.SNYK_EXCLUSIONS }}
 
     steps:
-      # Using Github checkout action to retrieve the codebase.
+      # Using Github checkout action to retrieve the pull request codebase.
```
Comment (NeilSchiefer) [inline .github/workflows/build-and-test.yml]: Isn't this workflow run in forked pull requests and regular branches?

**Review by motionsuggests — DISMISSED** [review]
💯

File: .github/workflows/build-and-test.yml (hunk 1 above)
Comment (eslutz) [reply .github/workflows/build-and-test.yml]: Yes, the workflow does run on all pull requests, but a few lines down, on line 28, we are specifically checking out the code tied to the pull request.  By changing the workflow trigger to `pull_request_target` it "runs in the context of the base of the pull request, rather than in the context of the merge commit, as the `pull_request` event does." ([events-that-trigger-workflows](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_target))

By checking out the PR code, it allows the code to be built and run unit tests and Snyk test on the updated code.  So it doesn't result in any difference for regular branches, but it allows the code from the forked repo pr to run and get tested.

Comment (NeilSchiefer) [reply .github/workflows/build-and-test.yml]: By adding the pr pull request specific stuff all the time, doesn't that mean that this will now *only* run for pull requests and not regular branches anymore?

Comment (eslutz) [reply .github/workflows/build-and-test.yml]: I don't think so, but I didn't test that out.  I'll give it a try and update once I find out.

**Comment by eslutz** [comment]
Snyk test now working running successfully on PRs from branches and forks.    I followed an example, linked below, from GitHub for being able to use a RockLib org token with code from a forked PR.  Since the snyk-test workflow isn't initiated by the pull request, I had to add code to provide the status of the workflow to the pull request for it to show as a check.  Additionally, in the process of figuring out how to get the results of the Snyk test into the pull request, I added an automation that leaves a comment in the pull request with the result of Snyk test and a link to the test (see screenshot below).  I thought it was cool so I left it, even though it's not needed since I got the status check working and updated.  If everyone thinks it would be better not to have the pull request comments I can remove that part of the code.

https://securitylab.github.com/research/github-actions-preventing-pwn-requests/

PR comments

File: .github/workflows/build-and-test.yml (hunk 1 above)
Comment (eslutz) [reply .github/workflows/build-and-test.yml]: Updates have been made to get the workflow running for both forked and branch PRs.

**Comment by eslutz** [comment]
Also, when I get back Tuesday, I'll be adding an issue and PR to update the process documentation for this.  I'll also be creating an issue for the updates that will have to be made to the other repos to get the PR testing workflows working again. 

No changes have been made to the publish workflow, so that still works as normal.

File: GenericWorkflows/build-and-test.yml [hunk 2]
```diff
@@ -1,20 +1,19 @@
...
 
 ####################################################################################################
 ##  WORKFLOW JOBS
 ####################################################################################################
 jobs:
-  # Calls the shared build-and-test workflow.
-  call_build_and_test:
-    name: Call build-and-test workflow
-    uses: RockLib/RockLib.Workflows/.github/workflows/build-and-test.yml@main
+  # Calls the shared unit-test workflow.
```
Comment (NeilSchiefer) [inline GenericWorkflows/build-and-test.yml:15]: Why was this renamed when the unit-test.yml still builds and tests?

Comment (eslutz) [reply GenericWorkflows/build-and-test.yml:15]: Since there's now two separate workflow files that run as part of PR testing I thought it made sense to use similar names with `unit-test` and `snyk-test` since that's the primary purpose of those two workflows.

**Review by aatchison14 — APPROVED** [review]

**Review by motionsuggests — APPROVED** [review]

File: GenericWorkflows/build-and-test.yml (hunk 2 above)
Comment (NeilSchiefer) [reply GenericWorkflows/build-and-test.yml:15]: Oh ok I thought it persisted the build for publishing.  But it rebuilds.  Carry on.

**Review by NeilSchiefer — APPROVED** [review]
